packages/core/saved-objects/core-saved-objects-utils-server/src/merge_migrations.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { ary, isFunction, mergeWith } from 'lodash';
+import type {
+  SavedObjectMigrationContext,
+  SavedObjectMigration,
+  SavedObjectMigrationMap,
+  SavedObjectUnsanitizedDoc,
+} from '@kbn/core-saved-objects-server';
+
+/**
+ * Composes two migrations into a single migration.
+ * ```
+ * mergeSavedObjectMigrations(outer, inner) -> (doc, context) => outer(inner(doc, context), context) }
+ * ```
+ *
+ * If at least one of the migrations is not deferred, the composed migration will not be deferred.
+ *
+ * @public
+ * @param outer Wrapping migration.
+ * @param inner Wrapped migration.
+ * @param rest Additional wrapped migrations to compose.
+ * @returns The composed migration can be either a function or an object depending on the input migrations.
+ */
+export function mergeSavedObjectMigrations(
+  outer: SavedObjectMigration,
+  inner: SavedObjectMigration,
+  ...rest: SavedObjectMigration[]
+): SavedObjectMigration {
+  if (rest.length) {
+    return mergeSavedObjectMigrations(
+      mergeSavedObjectMigrations(outer, inner),
+      ...(rest as [SavedObjectMigration, ...SavedObjectMigration[]])
+    );
+  }
+  if (!inner || !outer) {
+    return inner || outer;
+  }
+
+  const merged = mergeWith(
+    { ...(isFunction(inner) ? { transform: inner } : inner) },
+    isFunction(outer) ? { transform: outer } : outer,
+    (innerValue, outerValue, key) => {
+      if (key === 'deferred') {
+        return !!(innerValue && outerValue);
+      }
+
+      if (key === 'transform') {
+        return (state: SavedObjectUnsanitizedDoc, context: SavedObjectMigrationContext) =>
+          outerValue(innerValue(state, context), context);
+      }
+
+      return inner ?? outer;
+    }
+  );
```

Comment:
NIT: personal opinion here, but 1/ I feel like this is overly complicated given what it does and 2/ it's not type safe and wouldnt properly detect new properties being added to the `SavedObjectMigrationParams` type, right? 

A (non fp/sexy, I know, but readable) way would be something like:

```ts
const mergeMigs(a: SavedObjectMigration, b: SavedObjectMigration): SavedObjectMigrationParams {
   const in: SavedObjectMigrationParams = isFunc(a) ? { transform: a, deferred: false } : a;
   const out: SavedObjectMigrationParams = isFunc(b) ? { transform: b, deferred: false } : b;
  return {
    deferred: in.deferred && b.deferred,
    transform: (doc: SavedObjectUnsanitizedDoc, context: SavedObjectMigrationContext) => 
       in.transform(out.transform(state, context), context);
   }
}
```

and then

```ts
const mergeSavedObjectMigrations = (migrations: SavedObjectMigration[]): SavedObjectMigrationParams => {
   // loop and call mergeMigs(prev, next)
}

Reply:
Refactored that 👍